ipn/ipnlocal/local.go

```diff
@@ -140,9 +141,23 @@ type LocalBackend struct {
 	statusChanged *sync.Cond
 }
 
+type clientGen func(controlclient.Options) (controlclient.Client, error)
+
 // NewLocalBackend returns a new LocalBackend that is ready to run,
 // but is not actually running.
 func NewLocalBackend(logf logger.Logf, logid string, store ipn.StateStore, e wgengine.Engine) (*LocalBackend, error) {
+	// TODO(apenwarr): change controlclient.New to return controlclient.Client?
+	// Then we could avoid this wrapper, at the expense of external tests
+	// having to typecast in the interface.
+	ccWrap := func(opts controlclient.Options) (controlclient.Client, error) {
+		return controlclient.New(opts)
+	}
+	return NewLocalBackendWithClientGen(logf, logid, store, e, ccWrap)
+}
+
+// NewLocalBackend returns a new LocalBackend that is ready to run,
+// but is not actually running.
+func NewLocalBackendWithClientGen(logf logger.Logf, logid string, store ipn.StateStore, e wgengine.Engine, ccGen clientGen) (*LocalBackend, error) {
```

Comment:
no exported API requiring unexported types. Legal in Go, but almost never what you want.

But really we've been actively getting rid of all these sorts of factory types like @danderson said.

If you need this, I'd instead add a method to LocalBackend to set the constructor/factory instead, but start with the normal one (ccWrap above).  Then ditch the clientGen type and just put it in the method signature, like:

```go
func (b *LocalBackend) SetControlClientGetterForTesting(newControlClient func(controlclient.Options) (controlclient.Client, error))
```

Reply:
I didn't want to guarantee that ccGen would not be called from inside NewLocalBackend(), and behaviour would be undefined if you set it more than once, so a setter isn't great; it really seems like a construction-time feature to me.

I'll leave it for y'all to clean up if you find some way to do it elegantly.